### PR TITLE
add pair_data_generator

### PR DESCRIPTION
**Pull request by faneshion**

### CODE REVIEW

File: matchzoo/data_generator/pair_data_generator.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+from matchzoo import reorganize_data_pack_pair_wise
+from .data_generator import DataGenerator
+
+
+class PairDataGenerator(DataGenerator):
+    """Generate pair-wise data...."""
+    def __init__(self, data_pack, num_dup: int=1, num_neg: int=1, **kwargs):
+        """:class:`PairDataGenerator` constructor...."""
+        self._num_dup = num_dup
```
Comment (uduse) [inline matchzoo/data_generator/pair_data_generator.py]: `num_dup` is not used again so just discard it.

File: matchzoo/data_generator/pair_data_generator.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        self._num_neg = num_neg
```
Comment (uduse) [inline matchzoo/data_generator/pair_data_generator.py]: `num_neg` is only used to calculate `steps` so just initialize a `self._steps` here.

File: matchzoo/data_generator/pair_data_generator.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    def num_instance(self) -> int:
+        """Get the total number of pairs...."""
+        return math.ceil(len(self._data_pack) / (self._num_neg + 1))
+
+    def _get_batch_of_transformed_samples(self, indices: np.array):
+        """Get a batch of paired instances...."""
+        paired_indices = []
+        steps = self._num_neg + 1
+        for index in indices:
+            paired_indices.extend(
```
Comment (uduse) [inline matchzoo/data_generator/pair_data_generator.py]: `extend` takes an iterable, so I guess you don't have to `list` it.

File: matchzoo/data_generator/pair_data_generator.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            positive sample.
+        """
+        self._steps = num_neg + 1
+        self._data_pack = self.reorganize_data_pack(data_pack,
+                                                    num_dup,
+                                                    num_neg)
+        # Here the super().__init_ must be after the self._data_pack
+        super().__init__(self._data_pack, **kwargs)
+
+    def reorganize_data_pack(self, data_pack: DataPack, num_dup: int = 1,
```
Comment (uduse) [inline matchzoo/data_generator/pair_data_generator.py]: Make it a `classmethod` if no `self` is used.

File: matchzoo/data_generator/pair_data_generator.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        self._num_neg = num_neg
+        self._data_pack = reorganize_data_pack_pair_wise(data_pack, num_dup,
+                                                         num_neg)
+        # Here the super().__init_ must be after the self._data_pack
+        super().__init__(self._data_pack, **kwargs)
+
+    @property
+    def num_instance(self) -> int:
+        """Get the total number of pairs...."""
+        return math.ceil(len(self._data_pack) / (self._num_neg + 1))
```
[the remark this answers is not in the document]
Comment (faneshion) [reply matchzoo/data_generator/pair_data_generator.py]: It is not the same as the base class.

**Pull request merged**